### PR TITLE
[WIP] Hidden attribute by Expression + hide with css or HTML

### PR DESCRIPTION
**Pull request by Shnoulle**
New feature #[13835](https://bugs.limesurvey.org/view.php?id=13835) : hidden attribute : by Expression Manager
New feature #[14748](https://bugs.limesurvey.org/view.php?id=14748): Really hidden equation question type (not in HTML)
New feature #[11665](https://bugs.limesurvey.org/view.php?id=11665): More ways to hide a question

### CODE REVIEW

**Comment by Shnoulle** [comment]
Need testing

**Comment by dominikvitt** [comment]
Please update your branch to the latest develop changes:
- Survey preview and execution don't show any questions, goes directly to the end message, so I can't test it
- Some files are included in this PR, but shouldn't be ( or if they should then some explanation is required ):
  -  application/third_party/test/sodium_compat-1.9.1.zip
  -  application/third_party/test/sodium_compat-1.9.1/.gitignore

Also, please change attribute caption 
- from "Hidden way" 
- to "Hide method" ( it would be more understandable for users )

and help text 
- from "Using completely and no HTML is produced for the question, else only a css class his used, this allow javascript workaround" 
- to "Use "Completely" to skip generating HTML for the question, use "Visually" to generate and hide question HTML via CSS which allows javascript workarounds  to be used"

**Comment by Shnoulle** [comment]
> Survey preview and execution don't show any questions, goes directly to the end message, so I can't test it

A survey withn hidden question ?

**Comment by dominikvitt** [comment]
No, I imported a survey with all possible questions, this survey works as expected on develop branch.
Branch update may help.

**Comment by Shnoulle** [comment]
OK :) and yes, surely since it was done when develop are broken :)

**Comment by Shnoulle** [comment]
For equation, it's not the same help :, move to : 

> 'Use "Completely" to skip generating HTML for the question, use "Visually" to use updated result of the equation in same page.'

**Comment by dominikvitt** [comment]
No problem, just adjust it accordingly

**Comment by dominikvitt** [comment]
I still can't see questions, probably your PR is making all questions hidden by default.
Questions HTML is not generated.

**Comment by Shnoulle** [comment]
Sorry, i move it to [WIP] time to control , i send you a [READY/checked] when I tested it (and send a sample lss in mantis)

**Comment by Shnoulle** [comment]
Arg  [14850: Always hide this question attribute not working](https://bugs.limesurvey.org/view.php?id=14850)

**Comment by dominikvitt** [comment]
There was problem with list radio, but it is fixed.
Did you pulled the latest version?

**Comment by Shnoulle** [comment]
Yes : https://bugs.limesurvey.org/view.php?id=14854

**Comment by Shnoulle** [comment]
Can't work on it https://bugs.limesurvey.org/view.php?id=14892

**Comment by maziminke** [comment]
@Shnoulle Can you explain how this new feature works? Does it offer options to select the way a question will be hidden?

**Comment by Shnoulle** [comment]
2 feature : 
1. 3 ways for the boolean : Yes, no and equation : https://github.com/LimeSurvey/LimeSurvey/pull/1258/files#diff-babbcaf24f65c1def398ab14b5214c30R557
2. Hidden by expression by var a previous page (checked only in PHP) : https://github.com/LimeSurvey/LimeSurvey/pull/1258/files#diff-babbcaf24f65c1def398ab14b5214c30R567
2. Hide in HTML or in css only (different default for equation and other), really **needed** for equation … (Set score in same page, a user with some idea can look at page to see his current score). https://github.com/LimeSurvey/LimeSurvey/pull/1258/files#diff-babbcaf24f65c1def398ab14b5214c30R579

Look at the related mantis please. And look at the help in current pull request … 😔 …
When it's ready (i mean : when develop can be used to do test …) : i send a sample survey in doc

File: application/helpers/questionHelper.php [hunk 1]
```diff
@@ -529,20 +529,71 @@ public static function getAttributesDefinitions()
...
+        'types'=>Question::QT_1_ARRAY_MULTISCALE.Question::QT_5_POINT_CHOICE.Question::QT_A_ARRAY_5_CHOICE_QUESTIONS.Question::QT_B_ARRAY_10_CHOICE_QUESTIONS.Question::QT_C_ARRAY_YES_UNCERTAIN_NO.Question::QT_D_DATE.Question::QT_E_ARRAY_OF_INC_SAME_DEC_QUESTIONS.Question::QT_F_ARRAY_FLEXIBLE_ROW.Question::QT_G_GENDER_DROPDOWN.Question::QT_H_ARRAY_FLEXIBLE_COLUMN.Question::QT_I_LANGUAGE.Question::QT_K_MULTIPLE_NUMERICAL_QUESTION.Question::QT_L_LIST_DROPDOWN.Question::QT_M_MULTIPLE_CHOICE.Question::QT_N_NUMERICAL.Question::QT_O_LIST_WITH_COMMENT.Question::QT_P_MULTIPLE_CHOICE_WITH_COMMENTS.Question::QT_Q_MULTIPLE_SHORT_TEXT.Question::QT_R_RANKING_STYLE.Question::QT_S_SHORT_FREE_TEXT.Question::QT_T_LONG_FREE_TEXT.Question::QT_U_HUGE_FREE_TEXT.Question::QT_X_BOILERPLATE_QUESTION.Question::QT_Y_YES_NO_RADIO.Question::QT_Z_LIST_RADIO_FLEXIBLE.Question::QT_EXCLAMATION_LIST_DROPDOWN.Question::QT_COLON_ARRAY_MULTI_FLEX_NUMBERS.Question::QT_SEMICOLON_ARRAY_MULTI_FLEX_TEXT.Question::QT_VERTICAL_FILE_UPLOAD.Question::QT_ASTERISK_EQUATION,
+            'category'=>gT('Display'),
+            'sortorder'=>102,
+            'inputtype'=>'text',
```
Comment (Shnoulle) [inline application/helpers/questionHelper.php]: Maybe a new input type : dropdown + text (with an option to activate text input)

**Comment by Shnoulle** [comment]
I close this one : 

1. [Hide in HTML or in css only](https://bugs.limesurvey.org/view.php?id=14748) : i make a new Pull requets for this one **(Needed only for Equation)**  
2. [New feature #13835](https://bugs.limesurvey.org/view.php?id=13835) : hidden attribute : by Expression Manager : **more easy and clean to do it in plugin only** :) (if needed : iu can put a core or demo plugin for this).

**Pull request closed without merging**